x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/FieldNameTranslators.java

```diff
@@ -35,75 +36,42 @@
-            if (fieldSortBuilder.getNestedSort() != null) {
-                throw new IllegalArgumentException("nested sorting is not supported for API Key query");
-            }
-            if (FieldSortBuilder.DOC_FIELD_NAME.equals(fieldSortBuilder.getFieldName())) {
-                searchSourceBuilder.sort(fieldSortBuilder);
-            } else {
-                final String translatedFieldName = translate(fieldSortBuilder.getFieldName());
-                fieldNameVisitor.accept(translatedFieldName);
-                if (translatedFieldName.equals(fieldSortBuilder.getFieldName())) {
-                    searchSourceBuilder.sort(fieldSortBuilder);
-                } else {
-                    final FieldSortBuilder translatedFieldSortBuilder = new FieldSortBuilder(translatedFieldName).order(
-                        fieldSortBuilder.order()
-                    )
-                        .missing(fieldSortBuilder.missing())
-                        .unmappedType(fieldSortBuilder.unmappedType())
-                        .setFormat(fieldSortBuilder.getFormat());
+    public static final FieldNameTranslators API_KEY_FIELD_NAME_TRANSLATORS = new FieldNameTranslators(
+        List.of(
+            new ExactFieldNameTranslator("creator.principal", "username"),
+            new ExactFieldNameTranslator("creator.realm", "realm_name"),
+            new ExactFieldNameTranslator("name", "name"),
+            new ExactFieldNameTranslator(API_KEY_TYPE_RUNTIME_MAPPING_FIELD, "type"),
+            new ExactFieldNameTranslator("creation_time", "creation"),
+            new ExactFieldNameTranslator("expiration_time", "expiration"),
+            new ExactFieldNameTranslator("api_key_invalidated", "invalidated"),
+            new ExactFieldNameTranslator("invalidation_time", "invalidation"),
+            // allows querying on any non-wildcard sub-fields under the "metadata." prefix
+            // also allows querying on the metadata field itself
+            new FlattenedFieldNameTranslator("metadata_flattened", "metadata")
+        )
+    );
 
-                    if (fieldSortBuilder.sortMode() != null) {
-                        translatedFieldSortBuilder.sortMode(fieldSortBuilder.sortMode());
-                    }
-                    if (fieldSortBuilder.getNestedSort() != null) {
-                        translatedFieldSortBuilder.setNestedSort(fieldSortBuilder.getNestedSort());
-                    }
-                    if (fieldSortBuilder.getNumericType() != null) {
-                        translatedFieldSortBuilder.setNumericType(fieldSortBuilder.getNumericType());
-                    }
-                    searchSourceBuilder.sort(translatedFieldSortBuilder);
-                }
-            }
-        });
+    public static final FieldNameTranslators USER_FIELD_NAME_TRANSLATORS = new FieldNameTranslators(
+        List.of(
+            new ExactFieldNameTranslator("username", "username"),
```

Comment:
nit: a static helper method for the case when query and index fields are the same might be useful for this and later for roles.